src/main/java/bio/terra/filesystem/FireStoreDao.java

```diff
@@ -0,0 +1,254 @@
+package bio.terra.filesystem;
+
+import bio.terra.filesystem.exception.FileSystemObjectNotFoundException;
+import bio.terra.metadata.Dataset;
+import bio.terra.metadata.FSDir;
+import bio.terra.metadata.FSFile;
+import bio.terra.metadata.FSObjectBase;
+import com.google.cloud.firestore.Firestore;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.function.Consumer;
+
+// One DAO to rule them all...
+//
+// Operations on a file often need to touch file, directory, and dependency collections;
+// that is, the FireStoreFileDao, the FireStoreDirectoryDao, and the FireStoreDirectoryDao.
+// The data to make an FSDir or FSFile is now spread between the file collection and the
+// directory collection, so a lookup needs to visit two places to generate a complete FSObject.
+// This class coordinates operations between the daos.
+//
+// There are several functions performed in this layer.
+//  1. Encapsulating the underlying daos
+//  2. Converting from dao objects into DR metadata objects
+//  3. Dealing with project, dataset, and snapshot objects, so the daos don't have to
+//
+@Component
+public class FireStoreDao {
+    private FireStoreDirectoryDao directoryDao;
+    private FireStoreFileDao fileDao;
+    private FireStoreUtils fireStoreUtils;
+
+    @Autowired
+    public FireStoreDao(FireStoreDirectoryDao directoryDao,
+                        FireStoreFileDao fileDao,
+                        FireStoreUtils fireStoreUtils) {
+        this.directoryDao = directoryDao;
+        this.fileDao = fileDao;
+        this.fireStoreUtils = fireStoreUtils;
+    }
+
+    public void createDirectoryEntry(Dataset dataset, FireStoreObject newObject) {
```

Comment:
Yes, there are inconsistency with the naming. I use the term object (bad choice) to refer to either files or directories in the directory dao. I use the term objectId to refer to those same things.

Maybe I should use the term `fileId` for all file system things: dirs and files. And call the data object FireStoreFile and FireStoreDirectoryEntry.

Now that I have all of these PR's stacked, I'm feeling resistant to making this change in DR-515. What do you think about adding another ticket for name cleanup? It would be quick to do as a follow on, I think.